gdal/ogr/ogrsf_frmts/shape/shpopen.c

```diff
@@ -504,21 +497,21 @@ SHPOpen( const char * pszLayer, const char * pszAccess )
 /*                                                                      */
 /*      Open the .shp and .shx files based on the basename of the       */
 /*      files or either file name.                                      */
+/*      There are 2 types of SHPOpen() : with .shx restoring code       */
+/*      (SHPOpenLLEx) and normal one.                                   */
```

Comment:
I have a suggestion to solve copy&paste problem this way:
Add a new function to Shapelib, like
int SHPAPI_CALL
      SHPRestoreSHX( const char *pszShapeFile, const char *pszAccess,
                  SAHooks *psHooks );

Then, change SHPOpenLLEx() prototype to this:
SHPOpenLLEx( const char *pszShapeFile, const char *pszAccess,
                  SAHooks *psHooks, int bRestoreSHX )

So, its declaration would be something like this:
```cpp
{
if ( !bRestoreSHX ) return SHPOpenLL()
else
{
if ( SHPRestoreSHX() ) return SHPOpenLL();
else { errormsg; }
}
}
```